webpack/webpack-workspace-resolve-plugin/src/WorkspaceLayoutCache.ts

```diff
@@ -0,0 +1,198 @@
+   */
+  public readonly contextLookup: LookupByPath<IResolveContext>;
+  /**
+   * A weak map of package JSON contents to their corresponding context objects
+   */
+  public readonly contextForPackage: WeakMap<object, IResolveContext>;
+
+  public readonly resolverPathSeparator: string;
+  public readonly normalizeToSlash: (input: string) => string;
+  public readonly normalizeToPlatform: (input: string) => string;
+
+  public constructor(options: IWorkspaceLayoutCacheOptions) {
+    const { workspaceRoot, cacheData, resolverPathSeparator = directorySeparator } = options;
+
+    if (resolverPathSeparator !== '/' && resolverPathSeparator !== '\\') {
+      throw new Error(`Unsupported directory separator: ${resolverPathSeparator}`);
+    }
+
+    const resolveContexts: IResolveContext[] = [];
+    const contextLookup: LookupByPath<IResolveContext> = new LookupByPath(undefined, resolverPathSeparator);
+
+    this.contextLookup = contextLookup;
+    this.contextForPackage = new WeakMap<object, IResolveContext>();
+
+    const normalizeToSlash: (path: string) => string =
+      resolverPathSeparator === '/' ? preservePath : backslashToSlash;
+    const normalizeToPlatform: (path: string) => string =
+      resolverPathSeparator === '/' ? preservePath : slashToBackslash;
+
+    this.resolverPathSeparator = resolverPathSeparator;
+    this.normalizeToSlash = normalizeToSlash;
+    this.normalizeToPlatform = normalizeToPlatform;
+
+    // Internal class due to coupling of deserialization.
+    class ResolveContext implements IResolveContext {
+      private readonly _serialized: ISerializedResolveContext;
+      private _descriptionFileRoot: string | undefined;
+      private _dependencies: LookupByPath<IResolveContext> | undefined;
+
+      public constructor(serialized: ISerializedResolveContext) {
+        this._serialized = serialized;
+        this._descriptionFileRoot = undefined;
+        this._dependencies = undefined;
+      }
+
+      public get descriptionFileRoot(): string {
+        if (!this._descriptionFileRoot) {
+          this._descriptionFileRoot = `${workspaceRoot}${resolverPathSeparator}${normalizeToPlatform(this._serialized.root)}`;
+        }
+        return this._descriptionFileRoot;
+      }
+
+      public findDependency(request: string): IPrefixMatch<IResolveContext> | undefined {
+        if (!this._dependencies) {
+          // Lazy initialize this object since most packages won't be requested.
+          const dependencies: LookupByPath<IResolveContext> = new LookupByPath(undefined, '/');
+          // Handle the self-reference scenario
+          dependencies.setItem(this._serialized.name, this);
+          for (const [key, ordinal] of Object.entries(this._serialized.deps)) {
+            dependencies.setItem(key, resolveContexts[ordinal]);
+          }
+          this._dependencies = dependencies;
+        }
+
+        return this._dependencies.findLongestPrefixMatch(request);
+      }
+    }
```

Comment:
Why is this class inline instead of at the root level of the file?

Reply:
Because its implementation is tightly coupled to local variables (`workspaceRoot`, `resolverPathSeparator`, `normalizeToPlatform`, `resolveContexts`) in the constructor of `WorkspaceLayoutCache`. It could be promoted, at the expense of having to add more fields to the class, but it isn't really worth the complexity compared to capturing the variables in closures.